components/child_process_monitor/child_process_monitor.cc

```diff
@@ -0,0 +1,190 @@
+/* Copyright (c) 2021 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/child_process_monitor/child_process_monitor.h"
+
+#if defined(OS_POSIX)
+#include <errno.h>
+#include <fcntl.h>
+#include <signal.h>
+#include <sys/types.h>
+#include <sys/wait.h>
+#include <unistd.h>
+#elif defined(OS_WIN)
+#include <windows.h>
+#endif
+
+#include <utility>
+
+#include "base/bind_post_task.h"
+#include "base/logging.h"
+#include "base/process/kill.h"
+#include "base/single_thread_task_runner.h"
+#include "base/task/task_traits.h"
+#include "base/task/thread_pool.h"
+#include "base/threading/sequenced_task_runner_handle.h"
+
+namespace brave {
+
+namespace {
+
+#if defined(OS_POSIX)
+int pipehack[2];
+
+void SIGCHLDHandler(int signo) {
+  int error = errno;
+  char ch = 0;
+  (void)write(pipehack[1], &ch, 1);
+  errno = error;
+}
+
+#if defined(OS_MAC)
+void SetupFD(int fd) {
+  int flags;
+  if ((flags = fcntl(fd, F_GETFD)) == -1)
+    VLOG(0) << "get fd flags errno:" << errno;
+  flags |= FD_CLOEXEC;
+  if (fcntl(fd, F_SETFD, flags) == -1)
+    VLOG(0) << "set fd flags errno:" << errno;
+}
+#endif
+
+void SetupPipeHack() {
+#if defined(OS_MAC)
+  if (pipe(pipehack) == -1)
+    VLOG(0) << "pipehack errno:" << errno;
+  SetupFD(pipehack[0]);
+  SetupFD(pipehack[1]);
+#else
+  if (pipe2(pipehack, O_CLOEXEC) == -1)
+    VLOG(0) << "pipehack errno:" << errno;
+#endif
+
+  int flags;
+  if ((flags = fcntl(pipehack[1], F_GETFL)) == -1)
+    VLOG(0) << "get flags errno:" << errno;
+  // Nonblock write end on SIGCHLD handler which will notify monitor thread
+  // by sending one byte to pipe whose read end is blocked and wait for
+  // SIGCHLD to arrives to avoid busy reading
+  flags |= O_NONBLOCK;
+  if (fcntl(pipehack[1], F_SETFL, flags) == -1)
+    VLOG(0) << "set flags errno:" << errno;
+
+  struct sigaction action;
```

Comment:
generally it's a good idea to set `SA_RESTART` in your sigaction flags, especially since your code is doing a lot of syscalls.

Reply:
fixed in 2ce89fab2bc58cb8ab04ef80cb0f2efaaa56cbbe